### PR TITLE
See a project's Defect Escape Rate for a given time period

### PR DESCRIPTION
**Pull request by LALOPSB**
## What does this PR do?
Adds a section on each project that displays Defect escape rate data, describing:
- A column chart describing the Defect Escape rate progression over the n previous weeks.
- A summary section showing:
  - The overall rate.
  - The total amount of bugs.
  - The amount of bugs, divided by environment.

## How does it work?
1. Once a day, a job is scheduled to retrieve all the information for bugs on jira projects
2. The information is persisted to our db on the `jira_issues` table.
3. The information is parsed and retrieved to the app as needed.

### Jira
- A new custom field should be set on jira issues to define the environment for the bugs. This value will represent the environment from which the bugs were reported. 

image

- Environment variables should be set: 
  - for credentials: `JIRA_ADMIN_USER`, `JIRA_ADMIN_PASSWORD`
  - for the custom field: `JIRA_ENVIRONMENT_FIELD`
  - for the root url: `JIRA_ROOT_URL` (This is useful for using the sandbox or different stages of testing)
  
- Jira projects records

For the section to be shown, the Jira projects must be associated manually to the projects themselves, since there's no direct automatic mapping between them.
This is done by adding a new record to the table like this:

```
JiraProject.create!(
  jira_project_key: 'EM',  # This is the key assigned from Jira to the project.
  project_name: 'Engineering metrics',  # This is the name for the project on Jira selected by the project owner.
  project: Project.find(1) # This is the project on our domain.
)
```

### Project with informed bugs

image

### Project with no informed bugs on the period

image

### Project with no associated jira board

image

## Link to story:

https://rootstrap.atlassian.net/browse/EM-217?atlOrigin=eyJpIjoiMDQxM2ZmZmYwZjMzNDcxMmE3NzAwMjVkN2QwNDE1MzgiLCJwIjoiaiJ9

### CODE REVIEW

File: app/models/jira_issue.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+# Indexes
+#
+#  index_jira_issues_on_jira_project_id  (jira_project_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (jira_project_id => jira_projects.id)
+#
+
+class JiraIssue < ApplicationRecord
```
Comment (santib) [inline app/models/jira_issue.rb:23]: no need to change it, but maybe `JIRA` could have been a namespace?

File: app/models/jira_issue.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  enum issue_types: { bug: 'bug', task: 'task', epic: 'epic', story: 'story' }
```
Comment (santib) [inline app/models/jira_issue.rb:24]: can't you use symbols here? like `{ bug: :bug, task: :task, epic: :epic, story: :story }`

File: app/models/project.rb [hunk 3]
```diff
@@ -49,6 +49,7 @@ class Project < ApplicationRecord
            source: :user
 
   has_one :code_climate_project_metric, dependent: :destroy
+  has_one :jira_project, dependent: :destroy
```
Comment (santib) [inline app/models/project.rb:52]: is it possible that there are N JIRAs for 1 project? for example a big project with multiple pods where each pod uses a different JIRA board?

File: app/views/development_metrics/projects.erb [hunk 4]
```diff
@@ -47,6 +47,22 @@
...
     </div>
+
+    <div class="metrics shadow-box">
+      <h4 class="p-3">Defect Escape Rate</h4>
+      <% if @defect_escape_rate %>
+        <div class="graph">
+          <%= render 'development_metrics/defect_escape_rate/main_metric', defect_escape_rate: @defect_escape_rate %>
+        </div>
+      <% else %>
+
```
Comment (santib) [inline app/views/development_metrics/projects.erb:58]:
```suggestion
```

File: app/views/development_metrics/projects.erb [hunk 5]
```diff
@@ -47,6 +47,22 @@
         <%= render 'development_metrics/pull_request_size/main_metric', pull_request_size: @pull_request_size %>
       </div>
     </div>
+
+    <div class="metrics shadow-box">
+      <h4 class="p-3">Defect Escape Rate</h4>
+      <% if @defect_escape_rate %>
```
Comment (santib) [inline app/views/development_metrics/projects.erb:53]: I think you are missing 2 spaces here

File: spec/services/processors/jira_project_defect_escape_rate_updater_spec.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+              value: 'production'
+            },
+            created: '2021-03-14T15:48:04.000-0300'
+          }
+        }
+      ]
+    end
+
+    before do
+      allow_any_instance_of(JiraClient::Repository).to receive(:bugs).and_return(bugs)
```
Comment (santib) [inline spec/services/processors/jira_project_defect_escape_rate_updater_spec.rb:23]: I think we have webmock to mock the external request here

**Review by santib — APPROVED** [review]
Great work! 👏

File: app/models/project.rb (hunk 3 above)
Comment (matteo95g) [reply app/models/project.rb:52]: for now, we are going to stick with a 1:1 relationship, we talked about this with @santiagovidal yesterday

File: app/models/jira_issue.rb (hunk 1 above)
Comment (matteo95g) [reply app/models/jira_issue.rb:23]: hey, @santib I found [this](https://stackoverflow.com/questions/40478054/rails-5-namespacing-models) answer in StackOverflow about namespacing models. What are your thoughts? Is it worth doing the refactor?

**Review by sebaherrera07 — APPROVED** [review]

File: app/services/processors/jira_project_defect_escape_rate_updater.rb [hunk 7]
```diff
@@ -8,13 +8,20 @@ def call
...
+        )
+
+        issue.update!(
           informed_at: bug_fields[:created],
-          environment: bug_fields[ENV['JIRA_ENVIRONMENT_FIELD'].to_sym][:value],
+          environment: environment_field ? environment_field[:value] : nil,
           issue_type: 'bug'
         )
+
+        issue.save!
```
Comment (santiagovidal) [inline app/services/processors/jira_project_defect_escape_rate_updater.rb:24]: Is `save!` necessary if you did an `update!` previously?

File: app/assets/stylesheets/metrics/_summary.scss [hunk 8]
```diff
@@ -31,4 +31,8 @@
       justify-content: space-between;
     }
   }
+
+  .extra-padding {
+    padding: 100px;
```
Comment (martinjaimem) [inline app/assets/stylesheets/metrics/_summary.scss:36]: what about using rem here?

File: app/models/jira_issue.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    development: :development,
+    qa: :qa,
+    staging: :staging,
+    production: :production
+  }
+
+  belongs_to :jira_project
+
+  validates :informed_at, presence: true
+  validates :issue_type, presence: true
```
Comment (martinjaimem) [inline app/models/jira_issue.rb:37]:
```suggestion
  validates :informed_at, :issue_type, presence: true
```

File: app/services/builders/chartkick/defect_escape_value_data.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
+module Builders
+  module Chartkick
+    class DefectEscapeValueData < Builders::Chartkick::ProjectData
+      def build_data(metrics)
+        values = metrics.inject({}) do |hash, metric|
+          hash.merge!(metric.value[:bugs_by_environment]) { |_key, old, new| new + old }
+        end
+        total = values.values.inject(:+)
+        total && rate = (values['production'].to_i + values['staging'].to_i) * 100 / total
```
Comment (martinjaimem) [inline app/services/builders/chartkick/defect_escape_value_data.rb:9]:
```suggestion
        rate = (values['production'].to_i + values['staging'].to_i) * 100 / total if total
```

File: app/services/builders/chartkick/development_metrics.rb [hunk 11]
```diff
@@ -35,11 +35,23 @@ class Project < DevelopmentMetrics
...
+          end
+          metrics
+        end
+
+        def defect_escape_rate_entities
+          { defect_escape_rate: %w[defect_escape_rate defect_escape_values] }
+        end
+
+        def project_has_jira_board_associated?(project_id)
+          ::Project.find(project_id).jira_project
```
Comment (martinjaimem) [inline app/services/builders/chartkick/development_metrics.rb:54]:
```suggestion
          ::Project.find(project_id).jira_project.present?
```

File: app/services/jira_client/base.rb [hunk 12]
```diff
@@ -0,0 +1,13 @@
+module JiraClient
+  class Base
+    private
+
+    def connection
+      Faraday.new(ENV['JIRA_ROOT_URL']) do |conn|
+        conn.basic_auth(ENV['JIRA_ADMIN_USER'], ENV['JIRA_ADMIN_PASSWORD'])
```
Comment (martinjaimem) [inline app/services/jira_client/base.rb:7]:
```suggestion
        conn.basic_auth(ENV.fetch('JIRA_ADMIN_USER'), ENV.fetch('JIRA_ADMIN_PASSWORD'))
```
not a must

File: app/services/processors/jira_project_defect_escape_rate_updater.rb (hunk 7 above)
Comment (martinjaimem) [reply app/services/processors/jira_project_defect_escape_rate_updater.rb:24]: Agree with @santiagovidal:
`update!` in fact calls `save!`

```
def update!(attributes)
      with_transaction_returning_status do
        assign_attributes(attributes)
        save!
      end
    end
 ```

File: spec/jobs/jira_defect_metrics_updater_job_spec.rb [hunk 13]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+describe JiraDefectMetricsUpdaterJob do
```
Comment (martinjaimem) [inline spec/jobs/jira_defect_metrics_updater_job_spec.rb:3]:
```suggestion
describe JiraDefectMetricsUpdaterJob do
```

I think this applies to all `_spec` files

File: spec/services/builders/chartkick/defect_escape_rate_data_spec.rb [hunk 14]
```diff
@@ -0,0 +1,41 @@
+require 'rails_helper'
+
+RSpec.describe Builders::Chartkick::DefectEscapeRateData do
+  subject { described_class.send(:new, 1, '') }
```
Comment (martinjaimem) [inline spec/services/builders/chartkick/defect_escape_rate_data_spec.rb:4]:
```suggestion
  subject { described_class.new(1, '') }
```
What about explicitly calling `new` ?

**Review by martinjaimem — COMMENTED** [review]
Not for this PR, but for later, maybe it's better not to make a PR with all the functionalities and code enhancements all at once, for example, we could have left the `Gemfile` out of this PR and make a separate PR for it.
Also, we could have left the views out of the scope of this PR, as the whole backend functionality can live without the `views` so that the PR gets merged before and reviewers don't have to go through all the files 👍 .

Having said that, thanks for enhancing some old files, those kinds of small things make the code much better 👏

**Review by martinjaimem — APPROVED** [review]

**Pull request merged**